README.md
# data-connector-ULM-positions
Data connector that calls the `getPositions()` function on the strategy's vault address. This is intended for UniLiquidiyManager type vaults.
The resulting string is an array of three arrays: lower ticks, upper ticks, and weights respectively. The positions are returned in this deconstructed format.

i.e. "[ [-12, -6], [12, 6], [1, 1] ]"

The genesis execution will have no positions to return:
i.e. "[ [], [], [] ]"

## Parameters
- isChainRead: boolean set to true to signal this data connector fetches on-chain data.

No other parameters are needed as the vault's address is injected into the config and used in forming the request.


CHANGELOG.md
## 1.0.0 (2023-03-15)


### Features

* :zap: ULM positions as-update ([7a54c13](https://github.com/SteerProtocol/data-connector-ULM-positions/commit/7a54c134293727088b07a7765d6123dcf4656ec2))
* MVP ([cb07c4a](https://github.com/SteerProtocol/data-connector-ULM-positions/commit/cb07c4a0a27aeb0e5df56bbc6d63e2e3925667eb))


### Bug Fixes

* :bug: Properties missing ([4cc18ac](https://github.com/SteerProtocol/data-connector-ULM-positions/commit/4cc18ac3d8b9aef8d00a78efb16a78caeffa4d97))

## 1.0.0 (2022-08-10)


### Features

* :rocket: Refactor to function specific calls ([967a7a4](https://github.com/SteerProtocol/data-connector-chain-call-assemblyscript/commit/967a7a48e86ac7e807040786eefe9c90a7bc886d))

## 1.0.0 (2022-06-14)


### Features

* :construction: Building out the bundle ([9c07481](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/9c074812190801c01b264a537f86534c47788317))
* Candles & Comments ([e6bf492](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/e6bf4926c5f962e63b88a7a575148eca30cdc13f))
* DataCollection Prototyping, callback design ([07ae6c1](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/07ae6c166b55e303c363b10309ee92cacff12fe4))


### Bug Fixes

* :bug: axios wiring ([f991de4](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/f991de45b20d1738318318fa80f7fce439c9321e))
* :bug: Debugging and simplifying ([d9749e0](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/d9749e018249ff5f9a6003b4331e1702a1e07213))
* :bulb: Clearing up comments for compilation ([710cab8](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/710cab8e6f1a5f3714033ff5ba7bf9908f9e570e))
* :fire: a good base point of old sys ([2764e54](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/2764e54af9054d43940b6fbc192e0402c0c154ae))
* :zap: asbuild:big conf? ([2719a63](https://github.com/SteerProtocol/data-connector-template-assemblyscript/commit/2719a6322a72ebf16983c8d91da2a5c56217e02e))


tests/utils.ts
export const input_config = `{
  "title": "Gets current positions from UniLiquidityManager Vault",
  "description": "Input config for reading the slot0 function on a Uniswapv3 pool",
  "type": "object",
  "required": [
    "isChainRead"
  ],
  "properties": {
    "isChainRead" : {
      "type": "boolean",
      "title": "Is this a view or pure contract call?",
      "default": true
    }
  }
}`

export const examplePositions = '[ [ 100500 ], [ 100680 ], [ 1 ] ]';

export const config = '{"address" : "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48","isChainRead" : true }'

export const response = `{
"abi":[{"inputs":[],"name":"getPositions","outputs":[{"internalType":"int24[]","name":"","type":"int24[]"},{"internalType":"int24[]","name":"","type":"int24[]"},{"internalType":"uint16[]","name":"","type":"uint16[]"}],"stateMutability":"view","type":"function"}],
"address":"0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
"arguments":[],
"method":"getPositions"
}`

tests/index.test.ts
import { config, examplePositions, input_config, response } from "./utils";
import fs from 'fs'
import {WasmModule, loadWasm} from "@steerprotocol/app-loader";

// We use untouched so that we can run the un-optimized version of the wasm which will provide better stacktraces
// const myModule = require("../untouchLoader");

function hexEncode(str: string): any {
  var hex, i;

  var result = "";
  for (i=0; i<str.length; i++) {
      hex = str.charCodeAt(i).toString(16);
      result += ("000"+hex).slice(-4);
  }

  return result
}

describe("WASM Transformation Module", () => {
  let myModule: WasmModule;

  beforeEach(async () => {
    myModule = await loadWasm(fs.readFileSync(__dirname + "/../build/debug.wasm"), {})
  });
  describe("Fetching ULM Data", () => {
    test("can return input config", async () => {
      // myModule = await loadWasm(fs.readFileSync(__dirname + "/../build/debug.wasm"), {})
      // Call the configForm function on the transformation bundle
      const result = myModule.config();
      // Check that the result is the same as the expected result
      // Fix some funky encoding
      let hexResult = hexEncode(result) as string;
      hexResult = hexResult.replace(/000d/g, '');
      const hexExpected = hexEncode(input_config);
  expect(hexResult).toEqual(hexExpected);
  });

    test("fails imporper config", async () => {
      const improperConfig = `{"config":"null"}`

      const timestamp = 1654012158
      // The actual strategy instantiation and execution
      expect(() => {
        myModule.initialize(improperConfig);
      }).toThrowError();
    });

    test("can return call obj", async () => {
      const timestamp = 1654012158
      // const _config = myModule.__pin(myModule.__newString(config));
      myModule.initialize(config);
      const result = myModule.execute("");
      let hexResult = hexEncode(result) as string;
      hexResult = hexResult.replace(/0002/g, '');
      hexResult = hexResult.replace(/0020/g, '');
      hexResult = hexResult.replace(/000d/g, '');
      let hexExpected = hexEncode(response);
      hexExpected = hexExpected.replace(/0002/g, '');
      expect(hexResult).toBe(hexExpected);
    });

    // test("can process a response and return a new config", async () => {
    //   const timestamp = 1654012158
    //   myModule.initialize(config, timestamp);
    //   const result = myModule.main(response_swaps);
    //   let hexResult = hexEncode(result) as string;
    //   hexResult = hexResult.replace(/000d/g, '');
    //   const hexExpected = hexEncode(secondCall);
    //   expect(hexResult).toBe(hexExpected);
    // });

    // test("can process the final response and return true for callback termination", async () => {
    //   const timestamp = 1654012158
    //   myModule.initialize(config, timestamp);
    //   myModule.main(response_swaps);
    //   const result = myModule.main(`{"data":{"swaps":[]}}`);
    //   expect(result).toBe("true");
    // });

    test("can run transformation and return positions", async () => {
      // const timestamp = 1654012158
      myModule.initialize(config);
      myModule.execute(examplePositions);
      const result = myModule.transform();
      let hexResult = hexEncode(result) as string;
      hexResult = hexResult.replace(/000d/g, '');
      const hexExpected = hexEncode(examplePositions);
      expect(hexResult).toBe(hexExpected);
    });

  });
});


assembly/index.ts
import { JSON } from "json-as/assembly";
import {  DataConnectorConfig, ExecutionContext  } from "@steerprotocol/strategy-utils/assembly";

// Data connector to read the total supply off a contract

// Local Variables
@serializable
class Config extends DataConnectorConfig{
  address: string = "";
}

  var positions: string = "";
  var address: string = "";

  // Initializes variables from the config file
  export function initialize(config: string): void {
    // parse through the config and assing locals
    const configJson: Config = JSON.parse<Config>(config);
    if (configJson.address == "" ||
      configJson.address == null) {
      throw new Error("Config not properly formatted");
    }
    address = configJson.address;
  }


  export function execute(response: string): string {
    
    // Check if respone is first call
    if (response == ''){
      // return payload
      return `{
        "abi" : [{"inputs":[],"name":"getPositions","outputs":[{"internalType":"int24[]","name":"","type":"int24[]"},{"internalType":"int24[]","name":"","type":"int24[]"},{"internalType":"uint16[]","name":"","type":"uint16[]"}],"stateMutability":"view","type":"function"}],
        "address" : \"` + address + `\",
        "arguments" : [],
        "method" : "getPositions"
      }`
    }

    // Here we will get the results as ints so we just pass it through
    // I.e. [ [ 100500 ], [ 100680 ], [ 1 ] ]
    positions = response;
    return 'true'
  } 

  // Here we are performing no transformation, so we ship the hex value back
  export function transform(): string {
    // resulting data will be array
    return positions;
  }

  // An example of what the config object will look like after being created via the configForm
  export function exampleInputConfig(): string {
    return `{"vaultAddress" : '0x8ad599c3A0ff1De082011EFDDc58f1908eb6e6D8', "isChainRead" : true}`
  }

  // Renders the config object in JSON Schema format, which is used
  // by the frontend to display input value options and validate user input.
  export function config(): string {
    return `{
  "title": "Gets current positions from UniLiquidityManager Vault",
  "description": "Input config for reading the slot0 function on a Uniswapv3 pool",
  "type": "object",
  "required": [
    "isChainRead"
  ],
  "properties": {
    "isChainRead" : {
      "type": "boolean",
      "title": "Is this a view or pure contract call?",
      "default": true
    }
  }
}`; 
  }
